types/luasql/postgres.d.tl

```diff
@@ -0,0 +1,26 @@
+local record sql
+    postgres: function():Driver
+    record Driver
+        connect:function(Driver,string):Connection
+    end
+    record Connection
+        setautocommit:function(Connection,boolean):boolean
+        escape: function(Connection,string):string
+        execute:function(Connection,string):Cursor|number
+        commit: function(Connection):boolean
+        rollback: function(Connection):boolean
+        close:function(Connection):boolean
+    end
+    enum ModeString
+        "n"
+        "s"
+    end
+    record Cursor
+        close:function(Cursor):boolean
+        fetch:function(Cursor,string,ModeString):{string|nil:string}
```

Comment:
I don' think the `|nil` here is doing anything, should be safe to remove (and IIRC I think we currenly have some limitations with unions as map keys, so it may actually cause problems down the line...)

Reply:
actually, looking at it I think it should be string|number. No idea why I put the |nil in that place

from the docs:
```
"n"
    the resulting table will have numerical indices (default)
"a"
    the resulting table will have alphanumerical indices 
```
http://keplerproject.github.io/luasql/manual.html

Reply:
ooh, I just thought of a _mean_ hack... 😈 which I just tested, and it actually works!

```
local enum Numeric
   "n"
end

local enum Alpha
   "s"
end

local record Cursor
   fetch: function(Numeric): {string}
   fetch: function(Alpha): {string:string}
end

local c: Cursor = {}

local a = c.fetch("n")

for _, x in ipairs(a) do -- a is an array!
   print(x)
end

local m = c.fetch("s")

local v = m["hello"]  -- m is a map!

print(v:upper())
```

Combining single-entry enums and function overloading in records, we can effectively simulate dependent-typing on string constants! 😱

Reply:
(if you switch the `"n"` and `"s"` in the example above, Teal _does_ report the expected compile errors!)

Reply:
wow.

Do you want me to do it this way or just return a {string|number:string} ?

Reply:
Thinking about it, I think the hack version actually has better UX for library users.

If you use `{string|number:string}`, then you need to cast with `as` when using the result as an array (e.g. for using the result with `ipairs`).

The single-string-enum hack would only cause annoyances (i.e. require casts) if someone uses a _variable_ in that argument instead of a literal, but I find that extremely unlikely.